File: c++/BalancedBinaryTree.cpp
class Solution {
public:
    int dfs(TreeNode* curNode) {
        if (curNode == NULL) {
            return 0;
        }
        
        int ldep = dfs(curNode->left);
        if (ldep < 0) {
            return ldep;
        }
        
        int rdep = dfs(curNode->right);
        if (rdep < 0) {
            return rdep;
        }
        
        if (abs(ldep - rdep) > 1) {
            return -1;
        }
        
        return max(ldep, rdep) + 1;
        
    }
    
    bool isBalanced(TreeNode* root) {
        if (root == NULL) {
            return true;
        }
        
        return (dfs(root) >= 0) ? true : false;
    }
};

File: python3/BalancedBinaryTree.py
class Solution:
    def isBalanced(self, root: TreeNode) -> bool:
        
        def dfs(curNode):
            if curNode == None:
                return (True, 0)
            
            left, ldep = dfs(curNode.left)
            if ldep < 0:
                return (False, -1)
            
            right, rdep = dfs(curNode.right)
            if rdep < 0:
                return (False, -1)
            
            if abs(ldep - rdep) > 1:
                return (False, -1)
                        
            return (True, max(ldep, rdep) + 1)
        
        return dfs(root)[0]
